pandas/core/arrays/arrow/array.py

```diff
@@ -120,19 +120,62 @@ def cast_for_truediv(
             # Intentionally not using arrow_array.cast because it could be a scalar
             # value in reflected case, and safe=False only added to
             # scalar cast in pyarrow 13.
-            return pc.cast(arrow_array, pa.float64(), safe=False)
-        return arrow_array
+            # In arrow, common type between integral and float64 is float64,
+            # but integral type is safe casted to float64, to mirror python
+            # and numpy, we want an unsafe cast, so we cast both operands to
+            # to float64 before invoking arrow.
+            return pc.cast(arrow_array, pa.float64(), safe=False), pc.cast(
+                pa_object, pa.float64(), safe=False
+            )
+
+        return arrow_array, pa_object
 
     def floordiv_compat(
         left: pa.ChunkedArray | pa.Array | pa.Scalar,
         right: pa.ChunkedArray | pa.Array | pa.Scalar,
     ) -> pa.ChunkedArray:
-        # Ensure int // int -> int mirroring Python/Numpy behavior
-        # as pc.floor(pc.divide_checked(int, int)) -> float
-        converted_left = cast_for_truediv(left, right)
-        result = pc.floor(pc.divide(converted_left, right))
         if pa.types.is_integer(left.type) and pa.types.is_integer(right.type):
+            # Use divide_checked to ensure cases like -9223372036854775808 // -1
+            # don't silently overflow.
+            divided = pc.divide_checked(left, right)
+            # If one operand is a signed integer, we need to ensure division
+            # rounds towards negative infinity instead of 0. If divided.type
+            # is a signed integer, that means at least one operand is signed
+            # otherwise divided.type would be unsigned integer.
+            if pa.types.is_signed_integer(divided.type):
+                # GH 56676: avoid storing intermediate calculating in
+                # floating point type.
+                has_remainder = pc.not_equal(pc.multiply(divided, right), left)
+                result = pc.if_else(
+                    pc.and_(
+                        has_remainder,
+                        pc.less(
+                            pc.bit_wise_xor(left, right),
```

Comment:
Not a deal breaker but instead of the bitwise xor can is it easy to just xor comparisons of each argument being < 0? I would be willing to sacrifice some optimization potential here for readability

Reply:
Would it help with readability if I extracted logic for pc.less(pc.bit_wise_xor(left, right), 0) into a variable named has_one_negative_operand? Since I don't think pc.and_ has any short-cutting logic since it takes values, not expressions. I think that name communicates the intent of the logic and keeps the optimization, but what you suggested, of doing logical bit wise xor of both operands less than 0 would work too. Let me know what you prefer.

Reply:
Ah great suggestion - yea I think that is a viable compromise

Reply:
Yes I can change this.

Reply:
If there is a way to scope that so it only evaluates when the operands are not unsigned that would be better as well. You have some very nice tricks here, but not every Python developer may know the impacts of bitwise operations like this. So just want to make sure it doesnt get refactored into something unintended in the future

Reply:
> If there is a way to scope that so it only evaluates when the operands are not unsigned that would be better as well.

I am not aware of this, pc.if_else from what I can tell, even if the condition is all true, or all false, evaluates both conditions, since it takes an array of values, but not an expression, so I'm not sure there's a way to avoid evaluating it for both sides. If there's not a way to do this, it may be a useful feature for arrow, as this seems like an optimization that would benefit many use cases.